### DISCUSSION TITLE
beforeAll/afterAll stack rather than list

### DISCUSSION THREAD

**Discussion by kristiandupont** [vitest-dev/vitest#1364] [General]
I created [this issue](https://github.com/facebook/jest/issues/12091) in the Jest repo a long time ago. I have now started migrating to Vitest which seems damn near perfect, but sadly this particular issue exists here as well. Basically, `before*/after*` calls seem to be ordered as a list rather than a stack.

In Jasmine, where they were stack-ordered, I was able to create [test "hooks"](https://kristiandupont.medium.com/test-hooks-be89b760d2db) --- small functions that encapsulate bits of setup/teardown code. This is, IMO, a very useful pattern that helps create reusable and composable test fixture code. With the current ordering, this is impossible.

I would be happy to try and help with changing this but I want to make sure that I wouldn't be actively working against anyone.

**Comment by kristiandupont** [comment]
I understand that this issue is not trivial to solve or even grasp. I am happy to look into it myself, I would love a bit of help as to where to look and to make sure this isn't deliberately going against the intended behavior.

**Comment by kristiandupont** [comment]
Case in point, this appears to be exactly how setup/teardown functions in global setup files are executed, probably for the same reasons.

<img width="698" alt="image" src="https://user-images.githubusercontent.com/847742/175307630-4982fe33-c6eb-4932-819a-3e62b9da7350.png">

**Comment by sheremet-va** [comment]
Hello 👋🏻 Sorry for the very late response. 

Currently these hooks are called in parallel. I don't have anything against this approach, and PR is always welcome. The implementation is here:

https://github.com/vitest-dev/vitest/blob/63dca9956a29e79a0796ea3e59f833d02ad3058f/packages/vitest/src/runtime/run.ts#L51

**Reply by kristiandupont** [reply]
Awesome, thank you!

I wonder if there is any reason someone might want to retain the existing behavior? Since execution order is undefined, it couldn't be that they are upset about it changing, but I guess it might have some performance implications. If so, should it be possible to configure?
